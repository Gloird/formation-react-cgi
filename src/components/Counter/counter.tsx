import React, { Component } from "react";
import Button from "../Button/button";

//Création d'un interface pour le state
interface CounterState {
  counter: number;
}

//ICI pas de props donc on met {}
class Counter extends Component<{}, CounterState> {
  //Ici pas de type/interface pour le props donc on met {}
  constructor(props: {}) {
    //Il est obligatoir d'appeler le super avec au moins les props
    super(props);
    //Initialisation du sate
    this.state = { counter: 0 };
  }

  render() {
    return (
      <div>
        <Button
          texte="+1"
          onClick={() => {
            //Il est important d'utiliser un callbakc pour metre a jour le stat,
            //car dans ce cas la l'utilisateur peux spam le bouton est le state
            //doit bien faire +1 ou -1 avec le nombre qui ce trouve dans le state a ce moment là
            this.setState((state) => {
              return { counter: state.counter + 1 };
            });
          }}
        />
        {this.state.counter}
        <Button
          texte="-1"
          onClick={() => {
            //Il est important d'utiliser un callbakc pour metre a jour le stat,
            //car dans ce cas la l'utilisateur peux spam le bouton est le state
            //doit bien faire +1 ou -1 avec le nombre qui ce trouve dans le state a ce moment là
            this.setState((state) => {
              return { counter: state.counter - 1 };
            });
          }}
        />
      </div>
    );
  }
}

export default Counter;
