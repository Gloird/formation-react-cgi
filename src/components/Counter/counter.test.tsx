import { fireEvent, render } from "@testing-library/react";
import Counter from "./counter";

describe("Counter", () => {
  test("fonctionnement", () => {
    const renderCounter = render(<Counter />);

    //les 2 boutons sont dans le bom
    expect(renderCounter.getAllByRole("button")).toHaveLength(2);
    const boutonPlus = renderCounter.getAllByRole("button")[0];
    const boutonMoins = renderCounter.getAllByRole("button")[1];
    //Que le bouton 1 a contien +1
    expect(boutonPlus).toHaveTextContent("+1");
    //Que le bouton 2 a contien -1
    expect(boutonMoins).toHaveTextContent("-1");
    //Que il y a bien 0 de présent dans le dom
    expect(renderCounter.getByText("0")).toBeInTheDocument();

    const compteur = renderCounter.getByText("0");

    //Test du bon fonctionnement
    //Clique sur le bouton +1
    fireEvent.click(boutonPlus);
    //Présence du 1 dans le dom
    expect(compteur).toHaveTextContent("1");

    //Clique sur le bouton -1
    fireEvent.click(boutonMoins);
    //Présence du 0 dans le dom
    expect(compteur).toHaveTextContent("0");

    //Clique sur le bouton -1
    fireEvent.click(boutonMoins);
    //Présence du 1 dans le dom
    expect(compteur).toHaveTextContent("-1");

    //Clique sur le bouton +1
    fireEvent.click(boutonPlus);
    //Présence du 0 dans le dom
    expect(compteur).toHaveTextContent("0");
  });

  /**
   * Snapshot du composant pour valider les évolutions.
   */
  test("snapshot pour le composant Counter", () => {
    const renderCounter = render(<Counter />);

    expect(renderCounter.baseElement).toMatchSnapshot();
  });
});
