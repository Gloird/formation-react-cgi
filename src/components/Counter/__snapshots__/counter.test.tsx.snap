// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter snapshot pour le composant Counter 1`] = `
<body>
  <div>
    <div>
      <button
        style="border: 1px solid;"
      >
        +1
      </button>
      0
      <button
        style="border: 1px solid;"
      >
        -1
      </button>
    </div>
  </div>
</body>
`;
